Add options for overriding default pkg-config directories

lantw44 opened the pull request:
Since https://github.com/mesonbuild/meson/pull/4410 changes the
installation directory of pkg-config files on FreeBSD, it is no longer
possible for a portable script, such as JHBuild, to use the same
directory layout on different platforms. Projects not using Meson's
pkgconfig module are likely to keep installing pkg-config files into
lib/pkgconfig, so we end up having both libdata/pkgconfig and
lib/pkgconfig on FreeBSD. Unless JHBuild wants to patch build files for
other build systems or manually move files after installation like what
FreeBSD ports currently do, we will have a mixed directory layout when
using a custom prefix on FreeBSD.

Instead of hard-coding the path of pkg-config files, provide options
for users to override the default values. JHBuild can just set
'--libpkgconfigdir lib/pkgconfig --sharepkgconfigdir share/pkgconfig'
to keep its existing behavior like what it already does to disable
Debian-style multiarch with '--libdir lib'. Projects not using Meson's
pkgconfig module are also encouraged to use these options to ensure a
consistent directory layout.

mensinda commented on mesonbuild/mintro.py:
```diff
@@ -189,12 +189,14 @@ def list_buildoptions(coredata: cdata.CoreData) -> List[Dict[str, Union[str, boo
                         'infodir',
                         'libdir',
                         'libexecdir',
+                        'libpkgconfigdir',
```
`dir_option_names` in mconf.py also has to be updated.

lantw44 replied:
The patch already includes `dir_option_names` change for mesonbuild/mconf.py, so I don't have to do anything for this comment, right?

mensinda replied:
Yes, this is resolved now.

lantw44 commented:
> Also, `sharepkgconfigdir` doesn't appear to affect anything. It is just declared in cordata with a default value and never actually read (unless I am missing something).

Yes, `sharepkgconfigdir` is never read because currently meson always uses `libpkgconfigdir`. The reason I include it is that pkg-config searches both `lib/pkgconfig` and `share/pkgconfig` by default. `share/pkgconfig` is usually used by projects which don't install libraries, such as adwaita-icon-theme and iso-codes.

mensinda commented:
OK, but (in general) we try to avoid adding new meson options where possible to keep the list manageable (for both us and the user). So adding an option that does nothing is not a good idea. It is confusing for everyone involved.

So, please either remove the `sharepkgconfigdir` option or use it in pkg-config module (maybe with a new kwarg to switch between the two).

xclaesse commented on docs/markdown/Builtin-options.md:
```diff
@@ -53,6 +55,10 @@ accurate, especially on Linux where different distributions have different
+Both `libpkgconfigdir` and `sharepkgconfigdir` default to `libdata/pkgconfig`
+on FreeBSD, and `<default libdir>/pkgconfig` and `share/pkgconfig` otherwise.
+"default libdir" means the default value of `libdir`.
+
```
That doesn't tell why/how to use those options.

As far as I understand this patch, libpkgconfigdir is used by the pkgconfig module as default install dif for .pc files. But I don't see where sharepkgconfigdir gets used.

lantw44 replied:
`libpkgconfigdir` is for architecture-dependent library files and `sharepkgconfigdir` is for architecture-independent data files.

xclaesse replied:
Ok, thanks for the clarification, that should be told in the doc ;-)

As far as I know we don't support architecture-independent generator yet, I remember there is a PR somewhere... or did that got merged already?

lantw44 replied:
I know meson doesn't support generating architecture-independent .pc files. My intention was to allow projects which currently use `configuration_data ` and `configure_file` to generate .pc files to have a way to handle the path properly.

mensinda replied:
#5775 should provide architecture-independent .pc files as far as I can tell.

lantw44 replied:
> Ok, thanks for the clarification, that should be told in the doc ;-)

```
| libpkgconfigdir                      | see below     | Pkg-config directory | 
| sharepkgconfigdir                    | see below     | Architecture-independent pkg-config directory |
```

I already add these two rows to the table in the same file. Do you think we need a more detailed explanation in the doc?

xclaesse replied:
I'm not sure we should have options about arch indep pc files, until we actually can generate them. I would leave sharepkgconfigdir outside the scope of the PR and add it in the other PR that adds dataonly pc file.

xclaesse commented:
I would have a single `pkg_config_install_dir` option that defaults to <libdir>/pkgconfig everywhere except on freeBSD where it would defaults to <prefix>/libdata/pkgconfig.

lantw44 commented:
> OK, but (in general) we try to avoid adding new meson options where possible to keep the list manageable (for both us and the user). So adding an option that does nothing is not a good idea. It is confusing for everyone involved.

I added it because pkg-config searches not only `lib/pkgconfig` but also `share/pkgconfig` by default. It is odd to support only one of them. If only `libpkgconfigdir` is provided, projects needing `share/pkgconfig` are likely to hard-code it as `<datadir>/pkgconfig`, causing problems for people requesting https://github.com/mesonbuild/meson/pull/4410 because FreeBSD maps both `lib/pkgconfig` and `share/pkgconfig` to `libdata/pkgconfig`.

> So, please either remove the `sharepkgconfigdir` option or use it in pkg-config module (maybe with a new kwarg to switch between the two).

Do you think it is good to call the new option `architecture_independent`? This option should be used to change the default value of `install_dir` from `<libpkgconfigdir>` to `<sharepkgconfigdir>`. I am not sure if it is good to have an option whose only function is to change a default value, but it is probably the easiest way to make meson use it.

> I would have a single `pkg_config_install_dir` option that defaults to `<libdir>/pkgconfig` everywhere except on freeBSD where it would defaults to `<prefix>/libdata/pkgconfig`.

This does not handle the case where the project wants to install its .pc files for non-library data into `<datadir>/pkgconfig`. It is very likely that the project will just hard-code the directory as `<datadir>/pkgconfig` without providing a way for users to override it. For example,  [gsettings-desktop-schemas uses `install_dir: join_paths(datadir, 'pkgconfig')`](https://gitlab.gnome.org/GNOME/gsettings-desktop-schemas/blob/2bb055b3c1b87e6db80d891bd788f1d4b9a62af3/meson.build#L41). It also references https://github.com/mesonbuild/meson/issues/2253, which has not been fixed. It also does not work for projects which install .pc files into both `<libdir>/pkgconfig` and `<datadir>/pkgconfig`. It seems that [systemd just introduces its own options](https://github.com/systemd/systemd/blob/1cb342447d92053ab1750e498b6591f3dd46f0f6/meson.build#L116) to allow changing directories for installing these two kinds of .pc files.

mensinda commented:
> > So, please either remove the `sharepkgconfigdir` option or use it in pkg-config module (maybe with a new kwarg to switch between the two).
> 
> Do you think it is good to call the new option `architecture_independent`? This option should be used to change the default value of `install_dir` from `<libpkgconfigdir>` to `<sharepkgconfigdir>`.

I would go with something like a boolean `shared` or `install_shared`. This way, the name establishes a more obvious connection to the `sharepkgconfigdir` option (and is shorter). Though in the end @jpakkane also has to agree with the name and for the new functionality. For the time being, I would recommend to just go with a name that feels right and change it later if required.

>  I am not sure if it is good to have an option whose only function is to change a default value, but it is probably the easiest way to make meson use it.

Yes, there is already `install_dir`. Ideally, we would also warn if both `install_dir` and either `sharepkgconfigdir` or `libpkgconfigdir` is set to a non-default path. This way, we could make debugging problems easier.

Depending on the situation, we could also think about deprecating `install_dir` if the new options prove to be sufficient (or at least make it clear that setting `install_dir` might not be the best idea). I don't really see a use case for someone to manually set it, because there are only relatively few places where you would reasonably put a `.pc` file. All this is assuming meson can do a good enough job here.

xclaesse commented:
I think `sharepkgconfigdir` should go with: https://github.com/mesonbuild/meson/pull/5775.

xclaesse commented on mesonbuild/mesonlib.py:
```diff
@@ -683,6 +683,18 @@ def default_libexecdir():
+    if is_freebsd():
+        return 'libdata/pkgconfig'
+    else:
+        return default_libdir() + '/pkgconfig'
```
This means that if I do `-Dlibdir=something` and I don't `-Dlibpkgconfigdir=something/pkgconfig` then my pc files won't end up in the right place. This is not only unfortunate, but also not backward compatible, as far as I can tell.

lantw44 replied:
Yes, it changes the behavior, and it was why I had to patch `run_project_tests.py` and `run_unittests.py` to pass tests. I don't know if meson has backward compatibility requirement, but I think it is confusing to have a default value of an option to depend on the user-specified value of another option. If we do so, users can no longer discover default values with `meson setup --help` command.

xclaesse replied:
Yes meson must keep backward compat when possible. This will break every packaging systems that set a different libdir. I'm not convinced we should allow arbitrary path for pkgconfig dirs here. It really feels like complicated way of saying: don't change dir for freebsd.

mensinda commented:
@xclaesse Agreed. I didn't know about this one.
@lantw44 In this case, please remove `sharepkgconfigdir` from this PR. Also, please add the warning for a case were both `install_dir` and `libpkgconfigdir` are set, to clearly communicate which directory is used.

xclaesse commented:
Correct me if I'm wrong, but ultimately what this PR is about is adding a way to disable the per-platform (only freebsd atm) pkgconfig install dir, right? Just thinking out loud here, but what about a simple boolean option instead? That would solve inconsistency you get if you do `-Dlibpkgconfigdir=lib/pkgconfig -Dlibdir=lib64`.

lantw44 commented:
> > > So, please either remove the `sharepkgconfigdir` option or use it in pkg-config module (maybe with a new kwarg to switch between the two).
> > 
> > 
> > Do you think it is good to call the new option `architecture_independent`? This option should be used to change the default value of `install_dir` from `<libpkgconfigdir>` to `<sharepkgconfigdir>`.
> 
> I would go with something like a boolean `shared` or `install_shared`. This way, the name establishes a more obvious connection to the `sharepkgconfigdir` option (and is shorter). Though in the end @jpakkane also has to agree with the name and for the new functionality. For the time being, I would recommend to just go with a name that feels right and change it later if required.

If we want a shorter name, I think `install_data` is probably better because the directory `share` is already called `datadir`. Calling it `shared` may be confused with shared libraries or `sharedstatedir`.

> > I am not sure if it is good to have an option whose only function is to change a default value, but it is probably the easiest way to make meson use it.
> 
> Yes, there is already `install_dir`. Ideally, we would also warn if both `install_dir` and either `sharepkgconfigdir` or `libpkgconfigdir` is set to a non-default path. This way, we could make debugging problems easier.

`install_dir` is set by the project and `libpkgconfigdir` is set by the user compiling the project. This means that whether the warning is triggered depends on not only the code itself but also options set by users. I think it can be confusing for both developers and users. If a user reports the warning to a developer without saying the option used for configuring the build, the developer probably can't reproduce the problem.

> Depending on the situation, we could also think about deprecating `install_dir` if the new options prove to be sufficient (or at least make it clear that setting `install_dir` might not be the best idea). I don't really see a use case for someone to manually set it, because there are only relatively few places where you would reasonably put a `.pc` file. All this is assuming meson can do a good enough job here.

I agree that `install_dir` may be unnecessary if meson can handle the installation path properly. Users should be able to control where a file is installed.

> I think `sharepkgconfigdir` should go with: #5775.

https://github.com/mesonbuild/meson/pull/5775 currently does not address the problem of `sharepkgconfigdir`. Did you mean I should propose it there, or add it after it was merged?

> @xclaesse Agreed. I didn't know about this one.
> @lantw44 In this case, please remove `sharepkgconfigdir` from this PR. Also, please add the warning for a case were both `install_dir` and `libpkgconfigdir` are set, to clearly communicate which directory is used.

Does it imply that it is expected for the `libpkgconfigdir` patch and the `sharepkgconfigdir` patch to be merged in the same release cycle?

> @lantw44 Correct me if I'm wrong, but ultimately what this PR is about is adding a way to disable the per-platform (only freebsd atm) pkgconfig install dir, right? Just thinking out loud here, but what about a simple boolean option instead? That would solve inconsistency you get if you do `-Dlibpkgconfigdir=lib/pkgconfig -Dlibdir=lib64`.

Since we are adding an option, I think it is better to add an option solving more problems than just a single use case. I don't know if there are other platforms having different conventions which are not known to meson. Providing an option means these differences can be supported without patching meson.

mensinda commented:
> #5775 currently does not address the problem of sharepkgconfigdir. Did you mean I should propose it there, or add it after it was merged?

Actually, now that I think about it no. #5775 is basically ready to merge as-is. Also, with the dev freeze today and the `0.53.0` release on the 29th, I think that it would be better to merge #5775 without `sharepkgconfigdir`. Then you can use the new `dataonly` after a rebase to switch between the lib and share default dir.

> Does it imply that it is expected for the libpkgconfigdir patch and the sharepkgconfigdir patch to be merged in the same release cycle?

Not necessarily. But seeing that the `0.53.0` dev freeze is today and the current open questions with this PR, I wouldn't expect these new options to land in `0.53.0`.

> install_dir is set by the project and libpkgconfigdir is set by the user compiling the project. This means that whether the warning is triggered depends on not only the code itself but also options set by users. I think it can be confusing for both developers and users. If a user reports the warning to a developer without saying the option used for configuring the build, the developer probably can't reproduce the problem.

Yes, but debugging why `libpkgconfigdir` doesn't do anything would be even more annoying for a user. The user and/or developer would have read the documentation for `instal_dir` (which is a lot to ask of a user) to even have a chance to figure out what is going on. So, silently ignoring a user setting will certainly lead to even more confusion.

To avoid as much confusion as possible the warning should also clearly state *why* it was triggered (`install_dir` and `libpkgconfigdir` differ), and maybe even have a link to the relevant documentation. Also, it should be clear which path will actually be used (`install_dir` in this case). I am aware that will lead to a quite verbose warning, but I don't see a better solution until/if we deprecate `install_dir`.

lantw44 commented:
> > #5775 currently does not address the problem of sharepkgconfigdir. Did you mean I should propose it there, or add it after it was merged?
> 
> Actually, now that I think about it no. #5775 is basically ready to merge as-is. Also, with the dev freeze today and the `0.53.0` release on the 29th, I think that it would be better to merge #5775 without `sharepkgconfigdir`. Then you can use the new `dataonly` after a rebase to switch between the lib and share default dir.

I think it is still a question that whether `dataonly` implies `sharepkgconfigdir`. If a project uses `dataonly` to prevent `Cflags` from being generated but wants to reference `libdir` in the .pc file, there should be a way to ask meson to install it in `libpkgconfigdir` instead. `evolution-data-server-1.2.pc` may be an example. It contains `privlibdir`, `privdatadir`, `privlibexecdir`, but it has neither `Cflags` nor `Libs`.

> > Does it imply that it is expected for the libpkgconfigdir patch and the sharepkgconfigdir patch to be merged in the same release cycle?
> 
> Not necessarily. But seeing that the `0.53.0` dev freeze is today and the current open questions with this PR, I wouldn't expect these new options to land in `0.53.0`.

OK, if 0.53.0 is mostly frozen, I will just revert https://github.com/mesonbuild/meson/pull/4410 for use in JHBuild. It is not a problem. It is easy to add a patch in JHBuild.
 
> > install_dir is set by the project and libpkgconfigdir is set by the user compiling the project. This means that whether the warning is triggered depends on not only the code itself but also options set by users. I think it can be confusing for both developers and users. If a user reports the warning to a developer without saying the option used for configuring the build, the developer probably can't reproduce the problem.
> 
> Yes, but debugging why `libpkgconfigdir` doesn't do anything would be even more annoying for a user. The user and/or developer would have read the documentation for `instal_dir` (which is a lot to ask of a user) to even have a chance to figure out what is going on. So, silently ignoring a user setting will certainly lead to even more confusion.
> 
> To avoid as much confusion as possible the warning should also clearly state _why_ it was triggered (`install_dir` and `libpkgconfigdir` differ), and maybe even have a link to the relevant documentation. Also, it should be clear which path will actually be used (`install_dir` in this case). I am aware that will lead to a quite verbose warning, but I don't see a better solution until/if we deprecate `install_dir`.

I think it may be easier to always warn when `install_dir` is different from the default value. This is not the best option because the default value of `libpkgconfigdir` is platform-dependent, but it may increase the possibility of seeing the warning during development because it can be triggered by testing on a different platform.

lantw44 commented:
I added a release note snippet.

lantw44 commented:
@mensinda I just found this section in the doc:

> Since 0.46 a `StaticLibrary` or `SharedLibrary` object can optionally be passed as first positional argument. If one is provided a default value will be provided for all required fields of the pc file:
>
> `install_dir` is set to `pkgconfig` folder in the same location than the provided library.
> `description` is set to the project's name followed by the library's name.
> `name` is set to the library's name.

Since the default `install_dir` set by meson has higher priority than `libpkgconfigdir`, this means setting `install_dir` on the library passed as the positional argument to `pkgconfig.generate` will causes the warning to be shown even if `install_dir` is not used on the `pkgconfig.generate` call.

@xclaesse Do you think we can stop providing a default value for `install_dir` to prevent `libpkgconfigdir` from being ignored?

lantw44 commented:
I updated the doc for pkgconfig module and added `install_data` keyword argument for switching between `libpkgconfigdir` and `sharepkgconfigdir`. Is it ready for review now?

mensinda commented:
Implementation wise, this looks mostly OK to me. Please use  `install_mode` instead of `install_data` and make it take a string instead of a boolean (possible values would be `'data'` and `'lib'` or something like that). Adding more options will be much easier this way.

I still have some conceptual questions:
- should we really warn if meson sets the default install dir from a library or should we just use the new options instead?
- what should happen if a project installs different libraries into different directories?
- should we deprecate `install_dir` (maybe even in this PR or the next release)?
- do we want to introduce a breaking change with regards to setting the `libdir` option?

I am personally in favor of deprecating `install_dir`, but I would like feedback from @xclaesse that (and the other points) first.

lantw44 commented:
> Please use `install_mode` instead of `install_data` and make it take a string instead of a boolean (possible values would be `'data'` and `'lib'` or something like that).

The name `install_mode` is already used by other functions to specify the file mode. Are you sure you want to reuse the name for a different meaning?

xclaesse commented on mesonbuild/modules/pkgconfig.py:
```diff
@@ -435,17 +437,32 @@ def parse_variable_list(stringlist):
+and the library, or set install_dir to get_option('{}') to override
+the default value which may comes from the library.
+
+Please see https://mesonbuild.com/Pkgconfig-module.html#pkggenerate for more details.
```
I don't think you can warn here, because there are legitimate cases where you want to put your pc file elsewhere. For example, GStreamer install pc files for plugins into `<libdir>/gstreamer-1.0/pkgconfig` for the case you want to static link to installed plugins. In that case user is responsible for setting PKG_CONFIG_PATH.

xclaesse commented:
tbh, I'm tempted to say we should revert https://github.com/mesonbuild/meson/pull/4410, if freebsd wants to do something different than everyone else, it's up to them to do it at packaging level. Which is probably what they do with every other build systems any way? How does it work with autotools/cmake? My general feeling is this is getting way too complicated for a niche case.

Also, this PR is about having same hierarchy on freebsd as on other platforms, but why is that even needed? As far as I understand those pc files won't be found on freebsd, so that's useless...

lantw44 commented:
> Which is probably what they do with every other build systems any way? How does it work with autotools/cmake?

As far as I know, it is done by [patching build files with `sed`](https://svnweb.freebsd.org/ports/head/Mk/Uses/pathfix.mk?view=markup&pathrev=521222) or [moving .pc files after they have been installed into `DESTDIR`](https://reviews.freebsd.org/D10129). CMake knows it has to [search `libdata/pkgconfig` for .pc files](https://gitlab.kitware.com/cmake/cmake/merge_requests/1108), but I don't think it knows how to install them.

> Also, this PR is about having same hierarchy on freebsd as on other platforms, but why is that even needed? As far as I understand those pc files won't be found on freebsd, so that's useless...

It is true that FreeBSD configures `pkgconf` with `--with-pkg-config-dir=${PREFIX}/libdata/pkgconfig:/usr/libdata/pkgconfig` where `PREFIX` defaults to `/usr/local`, so .pc files in `/usr/local/lib/pkgconfig` or `/usr/local/share/pkgconfig` won't be found by `pkgconf`. The use case of installing .pc files into `lib/pkgconfig` and `share/pkgconfig` is for a custom prefix, such as a JHBuild development environment. In a custon prefix, you must set `PKG_CONFIG_PATH` for `pkgconf` to find .pc files, so it doesn't matter if `pkgconf` can find it by default.

I proposed adding options after https://github.com/mesonbuild/meson/pull/4410 was merged because I think it is similar to the situation of `libdir` on Debian. On Debian, `libdir` defaults to `lib/x86_64-linux-gnu`, but you can set it to `lib` if you want the same layout as other systems in your custom prefix.

xclaesse commented:
> > Which is probably what they do with every other build systems any way? How does it work with autotools/cmake?
> 
> As far as I know, it is done by [patching build files with `sed`](https://svnweb.freebsd.org/ports/head/Mk/Uses/pathfix.mk?view=markup&pathrev=521222) or [moving .pc files after they have been installed into `DESTDIR`](https://reviews.freebsd.org/D10129). CMake knows it has to [search `libdata/pkgconfig` for .pc files](https://gitlab.kitware.com/cmake/cmake/merge_requests/1108), but I don't think it knows how to install them.

Ok, so freeBSD already does that at packaging level, so I'm not sure why we did it in meson if the first place.

> > Also, this PR is about having same hierarchy on freebsd as on other platforms, but why is that even needed? As far as I understand those pc files won't be found on freebsd, so that's useless...
> 
> It is true that FreeBSD configures `pkgconf` with `--with-pkg-config-dir=${PREFIX}/libdata/pkgconfig:/usr/libdata/pkgconfig` where `PREFIX` defaults to `/usr/local`, so .pc files in `/usr/local/lib/pkgconfig` or `/usr/local/share/pkgconfig` won't be found by `pkgconf`. The use case of installing .pc files into `lib/pkgconfig` and `share/pkgconfig` is for a custom prefix, such as a JHBuild development environment. In a custon prefix, you must set `PKG_CONFIG_PATH` for `pkgconf` to find .pc files, so it doesn't matter if `pkgconf` can find it by default.

Tools like jhbuild already have to set PKG_CONFIG_PATH, so why can't it set it to freebsd locations when run on freebsd? Again, I'm not sure why meson should get involved at all.

> I proposed adding options after #4410 was merged because I think it is similar to the situation of `libdir` on Debian. On Debian, `libdir` defaults to `lib/x86_64-linux-gnu`, but you can set it to `lib` if you want the same layout as other systems in your custom prefix.

That sounds like an hypothetical case that doesn't really happen in practice. Why would you place at any other place than libdir/pkgconfig unless it's on freebsd?

As far as I understand, our options are:
- Revert https://github.com/mesonbuild/meson/pull/4410 and leave it to downstream to ugly patch like they do with other build systems.
- Leave it unchanged, and leave it to tools like jhbuild to deal with platform differences by setting PKG_CONFIG_PATH accordingly.
- Add boolean option to disable the freebsd special case for tools like jhbuild.
- Merge this PR to leave full control to the user where pc files are installed.
  * backward compatibility issues, all debian packages needs to be patched because `-Dlibdir=lib/x86_64-linux-gnu` won't place pc files there any more without adding `-Dlibpkgconfigdir`
  * inconsistency if you do `-Dlibpkgconfigdir=lib/pkgconfig -Dlibdir=lib64.`
  * 2 more built-in options, one nore kwarg on pkg.generate() method
  * unclear how to handle when passing `install_dir` kwarg to generate().

IMHO, we should go with option 1 because it has not been released yet (right?), and it's always possible to revisit this whole issue for next release. 0.53.0 is due tomorrow. After the release, I would recommend going with option 2 because it seems reasonable to support freebsd specific dirs, but jhbuild case is too niche to justify the extra complexity of option 3, and they can easily special case freebsd too, as far as I know.

Peter2121 commented:
Sorry to be a little bit late joining the discussion. I am one of FreeBSD users, asked for #4410.

IMHO, #4410 should be included in meson ASAP as **it fixes the bug**. A software developer who creates a lib with pc files, using meson, entrust to meson to install these pc files so they will be usable (findable) by another projects, using the library. As for now, on FreeBSD meson **fails** to do it - the pc files installed by meson are **not** findable as they go to incorrect place. It **is** a bug, as the result is not the one, asked by a user (install pc files to a place where they can be find).

I perfectly understand the problem of JHBuild devs, but I think that this problem should be solved at the level of JHBuild. And their proposal seems to be _really_ interesting, as the possibility to configure the destination for pc files at the moment of building software can help in many special cases. For example, on Mac OS X the `brew` system normally installs pc files in non-standard places (on my system I have both `/usr/local/Cellar/openssl/1.0.2o_1/lib/pkgconfig/openssl.pc` and `/usr/local/Cellar/openssl@1.1/1.1.1d/lib/pkgconfig/openssl.pc`). I can imagine another scenarios when it would be interesting to put pc files in a 'special' places. So if meson could correctly manage these cases - it makes meson system definitely better.

So, the 4th option seems to be the best one for me. As about the Debian compatibility - I think that it could be possible to use the path, provided as `-Dlibdir` (if it is present) and not `default_libdir()` in mesonlib.py:690 - it solves the compatibility problem. As about the inconsistency between `-Dlibpkgconfigdir` and `-Dlibdir` - it is not a problem at all, if a developer asks to use `-Dlibpkgconfigdir` - probably he knows what he is doing. This option **must** override the destination of pc files in all cases. If someone wants to put a library to `<prefix>/lib64` and pc files to `<prefix>/lib/pkgconfig` - just let him do it! ;)

eli-schwartz commented:
Maybe meson could refrain from using "libdata" if the --prefix option points outside of /usr?

Peter2121 commented:
...BTW, I've just checked - FreeBSD is not the only OS where `pkgconfig` directory should go to `/usr/local/libdata` - there is at least DragonFlyBSD with the same rule.
```
# uname -s
DragonFly 
```

xclaesse commented:
> IMHO, #4410 should be included in meson ASAP as **it fixes the bug**. A software developer who creates a lib with pc files, using meson, entrust to meson to install these pc files so they will be usable (findable) by another projects, using the library. As for now, on FreeBSD meson **fails** to do it - the pc files installed by meson are **not** findable as they go to incorrect place. It **is** a bug, as the result is not the one, asked by a user (install pc files to a place where they can be find).

I agree it's nicer to respect platform specific layout. I'm suggesting to revert the commit just to take a step back and be sure to have all options on the table for next release. After the release it's too late.

Since you seems to know better FreeBSD ecosystem, how does other build systems deals with this?

> I perfectly understand the problem of JHBuild devs, but I think that this problem should be solved at the level of JHBuild.

I'm also tempted to leave it to JHBuild to set PKG_CONFIG_PATH to the appropriate per-platform location. That doesn't sounds difficult to do, but I would like to hear from them if there are anything preventing them from doing that.

> And their proposal seems to be _really_ interesting, as the possibility to configure the destination for pc files at the moment of building software can help in many special cases. For example, on Mac OS X the `brew` system normally installs pc files in non-standard places (on my system I have both `/usr/local/Cellar/openssl/1.0.2o_1/lib/pkgconfig/openssl.pc` and `/usr/local/Cellar/openssl@1.1/1.1.1d/lib/pkgconfig/openssl.pc`). I can imagine another scenarios when it would be interesting to put pc files in a 'special' places. So if meson could correctly manage these cases - it makes meson system definitely better.

If they build with `/usr/local/Cellar/openssl@1.1/1.1.1d/` as prefix, that's standard location. As far as I know we never got complains from anyone else than FreeBSD regarding location of pc files. I personally don't like adding complexity for hypothetical use cases, would be nice to hear from a brew developer what's are their requirements.
 
> So, the 4th option seems to be the best one for me. As about the Debian compatibility - I think that it could be possible to use the path, provided as `-Dlibdir` (if it is present) and not `default_libdir()` in mesonlib.py:690 - it solves the compatibility problem. As about the inconsistency between `-Dlibpkgconfigdir` and `-Dlibdir` - it is not a problem at all, if a developer asks to use `-Dlibpkgconfigdir` - probably he knows what he is doing. This option **must** override the destination of pc files in all cases. If someone wants to put a library to `<prefix>/lib64` and pc files to `<prefix>/lib/pkgconfig` - just let him do it! ;)

If I understand correctly, that's what I suggested in my review comments above, but then it means the default value for an option depends on user set value of another option. It's a bit special, I don't think any other option works like that. I'm not saying it's impossible, but it makes this whole thing more complex than it first look like.

Peter2121 commented:
> how does other build systems deals with this
As far as I know, with autotools 'it just works' (I've never had problems with manually installed software built using autotools), with CMake the devs should manage it manually in `CMakeLists.txt` (there are necessary options for it), so it could be OK for one software and not OK for another one.

> If they build with /usr/local/Cellar/openssl@1.1/1.1.1d/ as prefix, that's standard location. As far as I know we never got complains from anyone else than FreeBSD regarding location of pc files.
Ah, yes, indeed, brew probably plays with prefix.
As I've just mentioned, DragonFlyBSD has the same directory layout than FreeBSD (I have one DragonFlyBSD installed, I've just checked it). And this OS is somehow desktop-oriented (there are a lot of activity around desktop software), so probably one day they will come to ask the same thing ;)
It seems that there is the same situation in OpenBSD (I need to check it to be sure, but I remember some noise about it).

> It means the default value for an option depends on user set value of another option. It's a bit special, I don't think any other option works like that. I'm not saying it's impossible, but it makes this whole thing more complex than it first look like.
Hmm, I agree with you that is looks a little bit complex. But if this approach permits to avoid patching of Debian packages - maybe it could be acceptable? As for me - it looks logic, so, understandable ;)

eli-schwartz commented:
>  As far as I know, with autotools 'it just works' (I've never had problems with manually installed software built using autotools), with CMake the devs should manage it manually in `CMakeLists.txt` (there are necessary options for it), so it could be OK for one software and not OK for another one.

autotools provides an expansion for `@libdir@`, and any autotools program I've seen uses this both as the install directory for *.so files, and for 
```
pkgconfigdir = $(libdir)/pkgconfig
pkgconfig_DATA = libmyproject.pc
```

CMake is no different, pkg-config files get installed manually via:
```
install(FILES "${CMAKE_CURRENT_SOURCE_DIR}/libmyproject.pc"
        DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig)
```

meson is the odd one out, because meson actually has a native function for handling the "pkg-config" artifact type, and it's not just a dumb file to be installed in a given directory via an install_file() function.

P.S. Please fix your quoting.

xclaesse commented:
@eli-schwartz as far as I understand your comment, it means autotools/cmake cannot support FreeBSD layout without downstream patches. So meson is no different, is it?

@Peter2121 DragonFlyBSD and OpenBSD could also be special cased like we do for FreeBSD if that's confirmed.  I'm pretty much convinced that per-platform location is the right thing to do, what I'm much less convinced is if we should make it configuration by the user. As we saw, supporting user configurable location is less obvious than initially thought, and seems only required for niche case like JHBuild that should be able to work around it pretty easily on their side, unless they have reasons why it's not possible? IMHO supporting the platform specific layout in JHBuild is the best thing they can do, because that's where *BSD devs will expect files to be.

eli-schwartz commented:
@xclaesse I'm quite sure this must be the case.

My understanding of the original PR is that it was intended to be "let's make meson the first build system to actually get this right".

lantw44 commented:
> Tools like jhbuild already have to set PKG_CONFIG_PATH, so why can't it set it to freebsd locations when run on freebsd? Again, I'm not sure why meson should get involved at all.

I didn't say JHBuild couldn't do it. I know it can, but it will make the file system layout more ugly and more confusing. Instead of having either the standard layout (`lib/pkgconfig` and `share/pkgconfig`) or FreeBSD layout (`libdata/pkgconfig`), now you have both of them in a single prefix. All of `lib/pkgconfig`, `share/pkgconfig`, `libdata/pkgconfig` have to exist in `PKG_CONFIG_PATH` because there are always projects which don't use meson or hard-code `lib/pkgconfig` and `share/pkgconfig` in `meson.build` files.

I don't think JHBuild can do `sed`-patching like what FreeBSD ports do. It creates unnecessary diffs in directories managed by VCS, causing JHBuild to rebuild the project even if there is no change when running it the second time. Moving files in `DESTDIR` may be easier, but I am not sure if it is the right thing to do.

> Maybe meson could refrain from using "libdata" if the --prefix option points outside of /usr?

I don't like having a special case on the path. While most users install ports or packages into `/usr/local`, it is possible to configure FreeBSD ports to use a different prefix by setting `LOCALBASE` and `PREFIX` variables. It is possible to choose a path without `/usr` in it. In fact, [porttools](https://www.freshports.org/ports-mgmt/porttools), the tools for testing ports, already do it when you run `port test`. For example, when you run it under `/usr/ports/shells/bash`, it may set `PREFIX` to `/tmp/bash-5.0.11`. Having a special case on prefix may surprise maintainers.

> ...BTW, I've just checked - FreeBSD is not the only OS where `pkgconfig` directory should go to `/usr/local/libdata` - there is at least DragonFlyBSD with the same rule.

This is possibly because DragonFlyBSD ports are based on FreeBSD ports. You can see `${OPSYS} == FreeBSD` and `${OPSYS} == DragonFly` checks in FreeBSD ports sometimes.

> It seems that there is the same situation in OpenBSD (I need to check it to be sure, but I remember some noise about it).

`find /usr/local -name '*.pc'` shows that OpenBSD uses the standard layout:
```
/usr/local/lib/pkgconfig/libnghttp2.pc
/usr/local/lib/pkgconfig/libcurl.pc
/usr/local/lib/pkgconfig/liblz4.pc
...
/usr/local/share/pkgconfig/shared-mime-info.pc
/usr/local/share/pkgconfig/gsettings-desktop-schemas.pc
/usr/local/share/pkgconfig/adwaita-icon-theme.pc
/usr/local/share/pkgconfig/iso-codes.pc
...
```

I can't find any of `/usr/libdata/pkgconfig` and `/usr/local/libdata/pkgconfig` on OpenBSD. It seems that `pkg-config` on OpenBSD is a custom Perl script. It begins with:
```perl
my @PKGPATH = qw(/usr/lib/pkgconfig
		 /usr/local/lib/pkgconfig
		 /usr/local/share/pkgconfig
		 /usr/X11R6/lib/pkgconfig
		 /usr/X11R6/share/pkgconfig);
```

I think it suggests OpenBSD uses the standard layout.

xclaesse commented:
> > Tools like jhbuild already have to set PKG_CONFIG_PATH, so why can't it set it to freebsd locations when run on freebsd? Again, I'm not sure why meson should get involved at all.
> 
> I didn't say JHBuild couldn't do it. I know it can, but it will make the file system layout more ugly and more confusing. Instead of having either the standard layout (`lib/pkgconfig` and `share/pkgconfig`) or FreeBSD layout (`libdata/pkgconfig`), now you have both of them in a single prefix. All of `lib/pkgconfig`, `share/pkgconfig`, `libdata/pkgconfig` have to exist in `PKG_CONFIG_PATH` because there are always projects which don't use meson or hard-code `lib/pkgconfig` and `share/pkgconfig` in `meson.build` files.

I know that's not helping you, but it sounds like a bug in every other build systems, but not meson. I'm not sure we should work around a cmake/autotools bug in meson... You can easily support both places by setting PKG_CONFIG_PATH and blame the inconsistency on FreeBSD and other build systems.

lantw44 commented:
> leave full control to the user where pc files are installed.

The reason I chose this way and made a patch was that I thought having builtin options could discourage developers from hard-coding them in each project. We have `mandir` and `infodir`, and *BSD can set `mandir` to `man` (instead of the default `share/man`). Some *BSD also set `infodir` to `info` (instead of the default `share/info`). If we have options for .pc files, changing the directory for .pc files can be as easy as changing as the directory for man and info pages. `infodir` option exists even if there is no code in meson using it. Sometims a platform can change the layout. [FreeBSD changes `infodir` from `info` to `share/info` one year ago.](https://svnweb.freebsd.org/changeset/ports/484628)

eli-schwartz commented:
> The reason I chose this way and made a patch was that I thought having builtin options could discourage developers from hard-coding them in each project.

But developers already should not hard-code them in each project, and as far as I know, the only current hardcoding that happens is people (like me) hardcoding `configure_file(configuration: values_dict, input: 'foo.pc.in', install_dir: 'share/pkgconfig')` **due to #5775 not being available**. (This will be fixed by merging in #5775 and having projects migrate to depend on this.)

The question here is whether there's general value in adding yet another option to meson, resulting in increased maintenance burden and cognitive overhead (as all code inevitably does), or whether this specifically is not needed.

xclaesse commented:
I definitely don't want to encourage people to start installing pkg-config files in other locations, FreeBSD is already annoying enough like that. Since no other build system support it any way, there is no real value in allowing it in meson.

Again, if it was just a simple option to let user set the path, I wouldn't mind. But as it turned out while reviewing the details of this PR, it's much more complex than that, so we have the balance the pros and cons. So far I don't see much in the pros category, other than jhbuild installed layout is "ugly". And even that, it's arguably other build systems to blame and not meson.

lantw44 commented:
> But developers already should not hard-code them in each project, and as far as I know, the only current hardcoding that happens is people (like me) hardcoding `configure_file(configuration: values_dict, input: 'foo.pc.in', install_dir: 'share/pkgconfig')` **due to #5775 not being available**. (This will be fixed by merging in #5775 and having projects migrate to depend on this.)

https://github.com/mesonbuild/meson/pull/5775 itself is not sufficient. You still need a keyword argument in `pkgconfig.generate` to allow switching between `lib/pkgconfig` and `share/pkgconfig`. I think `dataonly` does not imply `share/pkgconfig`.

> * Leave it unchanged, and leave it to tools like jhbuild to deal with platform differences by setting PKG_CONFIG_PATH accordingly.

If we really choose this way, it means we will have more test failiures on FreeBSD because `lib/pkgconfig` is hard-coded in the test as well.
```
Failed test during validate: 'test cases/common/47 pkgconfig-gen'
Reason: Expected file usr/lib/pkgconfig/simple.pc missing.
Expected file usr/lib/pkgconfig/libfoo.pc missing.
Expected file usr/lib/pkgconfig/libhello.pc missing.
Extra file usr/libdata/pkgconfig/libfoo.pc found.
Extra file usr/libdata/pkgconfig/simple.pc found.
Extra file usr/libdata/pkgconfig/libhello.pc found.

Install dir contents:
  - usr/include/simple.h  - usr/libdata/pkgconfig/libfoo.pc  - usr/libdata/pkgconfig/simple.pc  - usr/libdata/pkgconfig/libhello.pc
```

eli-schwartz commented:
That sounds like a bizarre test case, shouldn't tests be testing that the project itself works, rather than testing that meson/ninja performed as documented?

eli-schwartz commented:
So.... You're saying the person who submitted the meson feature didn't update the meson testsuite to match the change?

lantw44 commented:
> So.... You're saying the person who submitted the meson feature didn't update the meson testsuite to match the change?

Yes, there is no CI configured for FreeBSD, so the change is not tested. I don't know how much effort is required to fix the test. There are a few `installed_files.txt` files hard-coding `lib/pkgconfig`.

dankegel commented:
I started the day cheerful and then I tried testing https://github.com/mesonbuild/meson/pull/7103 on freebsd.
